course8/course_en.md

```diff
@@ -0,0 +1,184 @@
+a[0] = 1
+a[1] = 2
+println(a) // Output: [1, 2, 0, 0, 0]
+```
+
+### Simple Implementation of Circular Queues
+
+```moonbit no-check
+struct Queue {
+  mut array: Array[Int]
+  mut start: Int
+  mut end: Int // end points to the empty position at the end of the queue
+  mut length: Int
+}
+
+fn push(self: Queue, t: Int) -> Queue {
+  self.array[self.end] = t
+  self.end = (self.end + 1) % self.array.length() // wrap around to the start of the array if beyond the end
+  self.length = self.length + 1
+  self
+}
+```
+
+### Expanding the Queue
+
+When the number of elements exceeds the length of the array, an expansion operation is required:
+
+```moonbit no-check
+fn push(self: Queue, t: Int) -> Queue {
+  if self.length == self.array.length() {
+    let new_array: Array[Int] = Array::make(self.array.length() * 2, 0)
+    let mut i = 0
+    while i < self.array.length(), i = i + 1 {
+      new_array[i] = self.array[(self.start + i) % self.array.length()]
+    }
+    self.start = 0
+    self.end = self.array.length()
+    self.array = new_array
+  }
+  self.push(t) // Recursive call to complete the element addition
+}
+```
+
+### Removing Elements
+
+```moonbit no-check
+fn pop(self: Queue) -> Queue {
+  self.array[self.start] = 0
+  self.start = (self.start + 1) % self.array.length()
+  self.length = self.length - 1
+  self
+}
+```
+
+### Maintaining the Length of the Queue
+
+```moonbit no-check
+fn length(self: Queue) -> Int {
+  self.length
+}
+```
+
+### Generic Version of Circular Queues
+
+```moonbit no-check
+fn make[T]() -> Queue[T] {
+  {
+    array: Array::make(5, T::default()), // Initialize the array with the default value of the type
+    start: 0,
+    end: 0,
+    length: 0
+  }
+}
+```
+
+## Implementation of Singly Linked Lists
+
+A singly linked list consists of a series of nodes, each containing data and a reference to the next node.
+
+### Definition of Nodes and Linked Lists
+
+```moonbit
+struct Node[T] {
+  val : T
+  mut next : Option[Node[T]]
+}
+
+struct LinkedList[T] {
+  mut head : Option[Node[T]]
+  mut tail : Option[Node[T]]
+}
+```
+
+### Adding Elements
+
+```moonbit
+fn push[T](self: LinkedList[T], value: T) -> LinkedList[T] {
+  let node = { val: value, next: None }
+  match self.tail {
+    None => {
+      self.head = Some(node)
+      self.tail = Some(node)
+    }
+    Some(n) => {
+      n.next = Some(node)
+      self.tail = Some(node)
+    }
+  }
+  self
+}
+```
+
+### Calculating the Length of the Linked List
+
+A simple recursive function is used to calculate the length of the linked list:
+
+```moonbit
+fn length[T](self : LinkedList[T]) -> Int {
+  fn aux(node : Option[Node[T]]) -> Int {
+    match node {
+      None => 0
+      Some(node) => 1 + aux(node.next)
+    }
+  }
+
+  aux(self.head)
+}
+```
+
+### Stack Overflow Problem
+
+When the linked list is too long, the recursive calculation of the length can cause a stack overflow. To solve this problem, we can use tail call optimization.
+
```

Comment:
There's demo in the course video. You should at least give some screenshots here.